### PR TITLE
Tell `subprocess.Popen` to disown the process. Fixes #475.

### LINKED ISSUE #475: Detach the child process in `papis open`
* papis version (`$ papis --version` or commit number): papis, version 0.12

# how to reproduce the issue
After opening a document within the terminal using
```
papis open <document_identifier>
```
the terminal cannot be closed as it waits for the sub-process, e.g. `xdg-open`, to finish.

# resolve the issue

## On Linux
Using the `nohup` tool and redirecting the `stdout` stream to `/dev/null` should resolve the issue.

## On Windows
Adding the following parameter to the `subprocess.Popen` constructor should resolve the issue.
```
creationflags=subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP
```

### PR DESCRIPTION
**Pull request by cheinigk**
Fixes the issue #475.

### CODE REVIEW

File: papis/utils.py [hunk 1]
```diff
@@ -71,9 +71,22 @@ def general_open(file_name: str,
...
+            "shell": False,
+            "stdin": None,
+            "stdout": None,
+            "stderr": None,
+            "close_fds": True
+        }
+        # Tell subprocess to detach the process.
+        if sys.platform == "win32":
+            popen_kwargs["creationflags"] = subprocess.DETACHED_PROCESS
+            popen_kwargs["creationflags"] |= subprocess.CREATE_NEW_PROCESS_GROUP
```
Comment (alexfikl) [inline papis/utils.py:84]: Why are these Windows only? Does it not apply to Unix shells?

File: papis/utils.py [hunk 2]
```diff
@@ -71,9 +71,22 @@ def general_open(file_name: str,
...
+        else:
+            cmd.insert(0, "nohup")
+            popen_kwargs["stdout"] = subprocess.DEVNULL
+            popen_kwargs["stderr"] = subprocess.DEVNULL
```
Comment (alexfikl) [inline papis/utils.py:88]: Why set these to `DEVNULL` only here?

**Review by alexfikl — COMMENTED** [review]
Thank you for the PR!
I had some questions on the special casing of the OS and then we can get it in.

File: papis/utils.py (hunk 1 above)
Comment (alexfikl) [reply papis/utils.py:84]: Ah, just noticed that these are Windows-only constants. Nevermind!

File: papis/utils.py [hunk 3]
```diff
@@ -71,9 +71,22 @@ def general_open(file_name: str,
...
-        subprocess.Popen(
-            cmd, shell=False,
-            stdin=None, stdout=None, stderr=None, close_fds=True)
+        popen_kwargs = {
+            "shell": False,
+            "stdin": None,
+            "stdout": None,
+            "stderr": None,
+            "close_fds": True
+        }
```
Comment (alexfikl) [inline papis/utils.py:80]:
```suggestion
        }	# type: Dict[str, Any]
```
You'll need to add that to make `mypy` happy.

File: papis/utils.py (hunk 1 above)
Comment (cheinigk) [reply papis/utils.py:84]: Exactly, no worries :-)

File: papis/utils.py (hunk 2 above)
Comment (cheinigk) [reply papis/utils.py:88]: I thought they are Unix only, but [the doc](https://docs.python.org/3/library/os.html#os.devnull) says, it's None on Windows. I'll put it to the default popen_kwargs.

**Comment by alexfikl** [comment]
Any idea why the test is failing? For some reason this call doesn't return an error code even though the command does not exist:
```
        result = self.invoke([
            "Krishnamurti",
            "--tool", "nonexistingcommand"
        ])
>       self.assertNotEqual(result.exit_code, 0)
```
Is it because of the `nohup`? Does that not return a proper error code for some reason?

**Comment by alexfikl** [comment]
> I think the problem is with the combination of `nohup` and running it in the background using `subprocess.Popen`. It's not possible to query it for a return code. The current situation is, that `Popen` raises an error if the command does not exist, but `nohup` exists on Linux. I've introduced an explicit check for the opener now.

Awesome! Thanks a lot for all your work here! I'll go ahead and merge this :rocket:

**Comment by cheinigk** [comment]
Thanks for your help with the PR and for merging!

**Pull request merged**